app/preload/database/transactionsDao.ts

```diff
@@ -161,45 +162,97 @@ async function hasPendingShieldedBalanceTransfer(fromAddress: string) {
-) {
-    const table = (await knex())(transactionTable);
-
+interface UpdatesAndAdditions<T> {
+    updates: T[];
+    additions: T[];
+}
+/**
+ * Find which of the given transactions already exists in the database.
+ * @returns updates, which are the transactions that already exist,
+ * and additions, which are those that don't already exist.
+ */
+async function findExistingTransactions<T extends Partial<TransferTransaction>>(
+    transactions: T[]
+): Promise<UpdatesAndAdditions<T>> {
+    const knexConnection = await knex();
     const hashes = transactions
         .map((t) => t.transactionHash || '')
         .filter((hash) => hash);
-    const existingTransactions: TransferTransaction[] = await table
+    let existingTransactions: TransferTransaction[] = await knexConnection
+        .table(transactionTable)
         .whereIn('transactionHash', hashes)
         .select();
 
-    const [updates, additions] = partition(transactions, (t) =>
+    const [updates, additionsOnHash] = partition(transactions, (t) =>
         existingTransactions.some(
             (t_) => t.transactionHash === t_.transactionHash
         )
     );
 
-    const additionChunks = chunkArray(additions, 50);
-    for (let i = 0; i < additionChunks.length; i += 1) {
-        // eslint-disable-next-line no-await-in-loop
-        await table.insert(additionChunks[i]);
+    if (!additionsOnHash.length) {
+        return { updates, additions: [] };
     }
 
-    await Promise.all(
-        updates.map(async (transaction) => {
-            const { transactionHash, ...otherFields } = transaction;
-            return updateTransaction(
-                { transactionHash: transaction.transactionHash },
-                otherFields
-            );
-        })
+    const ids = additionsOnHash.map((t) => t.id || '').filter((id) => id);
+    existingTransactions = await knexConnection
+        .table(transactionTable)
+        .whereIn('id', ids)
+        .select();
+
+    const [alsoUpdates, additions] = partition(additionsOnHash, (t) =>
+        existingTransactions.some((t_) => t.id === t_.id)
     );
 
+    return { updates: updates.concat(alsoUpdates), additions };
+}
+
+/**
+ * Upserts the provided list of transactions into the database as part of a database transaction.
+ * @param updates the array of transactions to update
+ * @param additions the array of transactions to insert
+ * @param trx knex transaction, which these upserts should be added to.
+ */
+async function upsertTransactionsTransactionally<
+    T extends Partial<TransferTransaction>
+>(updates: T[], additions: T[], trx: Knex.Transaction): Promise<void> {
+    const additionChunks = chunkArray(additions, 50);
+
+    for (const additionChunk of additionChunks) {
+        await trx.table(transactionTable).insert(additionChunk);
+    }
+
+    for (const updatedTransaction of updates) {
+        const { transactionHash, ...otherFields } = updatedTransaction;
+        if (transactionHash) {
```

Comment:
If `findExistingTransactions` returned `{ hashUpdates, idUpdates, additions }`, then this condition could be removed.

Reply:
Yes, I refactored it.